### ISSUE TITLE
Proposed change to events.lua

### ISSUE THREAD

**Issue by originalfoo** [Nexela/CircuitAlerter#2]
I have a proposed change to events.lua, but it's breaking change which is why I've not sent a PR yet:

https://github.com/aubergine10/CircuitAlerter/blob/master/lib/events.lua

The main changes are:
- Event data objects format has changed
  - Send full entity for "entity" scenario - although not sure what happens to it if it's no longer valid
  - Only player index is required for the "self" scenario
- Check for changes twice per second (`%30`), rather than 6 times per second (`%10`)
  - Considering we're tracking user actions, I don't think the extra 4 checks per second bring any benefit
- Two events will be triggered if entity A was closed due to entity B being opened
  - Removes need for external script to track state, it can be fully-event driven in all scenarios
- Cache of previous state has been altered - two separate props `opened` and `opened_self`

There is another reason why I've not sent a PR.... IMO it would be useful to put these events in a separate utility/library mod. Other mods can then hook in to the events (getting their ids from remote interface), without needing to be reliant on the CircuitAlerter mod. The opened/closed events are sorely lacking from Factorio API and it seems Factorio dev team is unwilling to implement any time soon, yet many mods need such events. The separate "Events" ("Eventorio"? lol) mod could gradually grow to provide a number of new events that modders can depend on.

**Comment by Nexela** [comment]
The basis for the events came from "Custom Events" mod on the portal but I did not want to rely on an external mod.  However I have no issue with shared libraries (i.e. factorio standard library) .

I do like your proposed changes though it is definitely a lot more clean and sane

I agree on the tick  changes, I actually forgot to reduce them when I reduced the other ones.

And yeah the devs need to not be stingy with open/closed events :)

**Comment by originalfoo** [comment]
Is there a reliable way to work out the parent GUI element of the dialog for the opened entity? If so, it would be worth sending that in the open event so event handlers can manipulate the GUI more easily.

**Comment by Nexela** [comment]
No way that I can see to get anything related to built-in entity gui's

**Comment by Nexela** [comment]
With your events when opened_self open/close should return player as the entity since the player is an entity

**Comment by originalfoo** [comment]
I was thinking of doing that but realised it was extra work for game engine to convert the Lua object to C and then back to Lua. Hence just sending `player.index` when open/close self. If the external event handler wants the player object, it can `game.players[ event.player_index ]`. It's a shame there's not some way (that I know of at least) to do similar with other types of entity - there doesn't seem to be any unique identifier for entities.

**Comment by originalfoo** [comment]
BTW... I've not done testing to check difference in performance between sending the player entity or not. I'm just assuming that sending just player index is faster, but I could be wrong. Doing a `game.players[ event.player_index ]` might actually be slower than just sending the player entity as part of event and have Factorio API do it's stuff.

**Comment by Nexela** [comment]
In the end it's sorta neglible as the speed difference gained or lost would
be the same as sending the entity when opening an entity versus the
player.  Add to that the event is raised only every 30 ticks.

**Comment by originalfoo** [comment]
Yup, good point, but the external script might not want to do anything with the player object (particularly when self closed). There's no performance hit by just sending the `player_index` should the external script need the full object, however there is a gain when the external script doesn't need the player object. I imagine quite a few mods will make use of the event, so even though it's rare event it's worth having that extra little optimisation IMO.

One other thing with the event working this way... we could ditch the `type` property from the event object, because the external script can easily work out what sort of event it is by presence of the `entity` property...

``` lua
function myEventHandler( event )
  if event.entity then
     -- entity was open/closed (player.open changed)
  else
     -- player was open/closed (player.open_self changed)
    local player = game.players[event.player_index] -- only if i need player object
  end
end
```
